src/advertisement/mod.rs

```diff
@@ -0,0 +1,148 @@
+use tracing::{debug, error};
+
+mod test;
+pub mod ticket;
+pub mod topic;
+
+#[derive(Debug)]
+pub struct AdNode {
+    node_record: Enr,
+    insert_time: Instant,
+}
+
+impl AdNode {
+    pub fn new(node_record: Enr, insert_time: Instant) -> Self {
+        AdNode {
+            node_record,
+            insert_time,
+        }
+    }
+}
+
+impl PartialEq for AdNode {
+    fn eq(&self, other: &Self) -> bool {
+        self.node_record == other.node_record
+    }
+}
+
+struct AdTopic {
+    topic: TopicHash,
+    insert_time: Instant,
+}
+
+impl AdTopic {
+    pub fn new(topic: TopicHash, insert_time: Instant) -> Self {
+        AdTopic { topic, insert_time }
+    }
+}
+
+pub struct Ads {
+    expirations: VecDeque<AdTopic>,
+    ads: HashMap<TopicHash, VecDeque<AdNode>>,
+    ad_lifetime: Duration,
+    max_ads_per_topic: usize,
+    max_ads: usize,
+}
+
+impl Ads {
+    pub fn new(
+        ad_lifetime: Duration,
+        max_ads_per_topic: usize,
+        max_ads: usize,
+    ) -> Result<Self, &'static str> {
+        if max_ads_per_topic > max_ads {
+            return Err("Ads per topic cannot be > max_ads");
+        }
+
+        Ok(Ads {
+            expirations: VecDeque::new(),
+            ads: HashMap::new(),
+            ad_lifetime,
+            max_ads_per_topic,
+            max_ads,
+        })
+    }
+
+    pub fn get_ad_nodes(&self, topic: TopicHash) -> impl Iterator<Item = Enr> + '_ {
+        self.ads
+            .get(&topic)
+            .into_iter()
+            .flat_map(|nodes| nodes.iter())
+            .map(|node| node.node_record.clone())
+    }
+
+    pub fn ticket_wait_time(&mut self, topic: TopicHash) -> Option<Duration> {
+        self.remove_expired();
+        let now = Instant::now();
+        if self.expirations.len() < self.max_ads {
+            self.ads
+                .get(&topic)
+                .filter(|nodes| nodes.len() >= self.max_ads_per_topic)
+                .map(|nodes| {
+                    nodes.get(0).map(|ad| {
+                        let elapsed_time = now.saturating_duration_since(ad.insert_time);
+                        self.ad_lifetime.saturating_sub(elapsed_time)
+                    })
+                })
+                .unwrap_or_default()
+        } else {
+            self.expirations.get(0).map(|ad| {
+                let elapsed_time = now.saturating_duration_since(ad.insert_time);
+                self.ad_lifetime.saturating_sub(elapsed_time)
+            })
+        }
+    }
+
+    fn remove_expired(&mut self) {
+        let mut map: HashMap<TopicHash, usize> = HashMap::new();
+
+        self.expirations
+            .iter()
+            .take_while(|ad| ad.insert_time.elapsed() >= self.ad_lifetime)
+            .for_each(|ad| {
+                let count = map.entry(ad.topic).or_default();
+                *count += 1;
+            });
+
+        map.into_iter().for_each(|(topic, index)| {
+            let entry_ref = self.ads.entry(topic).or_default();
```

Comment:
what I mean is that `or_default` should never be called. On one side, if we are removing things, we shouldn't be inserting keys with default values in general. And in this specific case, the key should exist given the way it was created. In practice, this code path will never be reached _as is_ right now. But leaving this here is prone to errors in future refactors.

TLDR: don't call `or_default` since here the purpose is not to ensure something is inserted